lib/pbench/server/api/resources/users_api.py

```diff
@@ -1,175 +1,70 @@
                         "username": <username>,
-                        "profile": <new_profile>
-                        ...
+                        "email": <email>
+                        "first_name": <firstName>,
+                        "last_name": <lastName>,
+                        "created": <registered_on>,
                     }
-            Failure: <status_Code>
+            Failure: <status_Code>,
                     response_object = {
                         "message": "failure message"
                     }
         """
-        target_user_id = params.uri["target_username"]
-        self._check_authorization(
-            ApiAuthorization(
-                ApiAuthorizationType.USER_ACCESS,
-                OperationCode.READ,
-                target_user_id,
-            )
-        )
-        target_user = User.query(id=int(target_user_id))
-        user_payload = params.body["profile"]
-
-        try:
-            valid_dict = target_user.form_valid_dict(**user_payload)
-            target_user.update(valid_dict)
-        except UserProfileBadKey as e:
-            current_app.logger.debug(
-                "User tried updating non updatable profile key, {}", user_payload
-            )
-            raise APIAbort(HTTPStatus.BAD_REQUEST, message=str(e))
-        except UserProfileBadStructure as e:
-            current_app.logger.debug("Bad user prfile structure, {}", user_payload)
-            raise APIAbort(HTTPStatus.BAD_REQUEST, message=str(e))
-        except Exception:
-            current_app.logger.exception(
-                "Exception occurred during updating user object"
-            )
-            raise APIInternalError(
-                HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR"
-            )
+        disabled = ServerConfig.get_disabled(readonly=True)
+        if disabled:
+            abort(HTTPStatus.SERVICE_UNAVAILABLE, **disabled)
+
+        auth_token = Auth.get_auth_token()
+        current_user = Auth.token_auth.current_user()
+        response_object = {}
+
+        if current_user.username == target_username:
+            token_payload = Auth.oidc_client.token_introspect(auth_token)
+            response_object = {
+                "username": token_payload.get("preferred_username"),
+                "first_name": token_payload.get("given_name"),
+                "last_name": token_payload.get("family_name"),
+                "email": token_payload.get("email"),
+            }
+        elif current_user.is_admin():
+            target_user = User.query(username=target_username)
+            if target_user:
+                response_object = target_user.as_json()
+            else:
+                abort(HTTPStatus.NOT_FOUND, message="Target user not found")
+        else:
+            abort(HTTPStatus.FORBIDDEN, message="Forbidden to perform the GET request")
 
-        response_object = target_user.get_json()
         return make_response(jsonify(response_object), HTTPStatus.OK)
```

Comment:
The reason to return the `last_name` `first_name` `email` fields from the token provided to the Server is that the dashboard could use this and populate these fields on the UI or store it in the cookie or something (right now it would have to make one more call to the OIDC userinfo endpoint to get this info or have to decode the token by itself). So that's why I am returning these entries when client presents their own token but if the request is from an admin user then we don't have to user token to decode the information all we can return is the cached entry of the target user.

Reply:
Why would the UI need `last_name`, `first_name`, or `email`?  (I don't think the Dashboard has any more use for those than the Server does....)

I understand why the returns to the user vs. to an Admin are different, but this feels like a case of doing something just because we can and not because we have any actual need for it.

Let's stick to doing what we need -- as long as we are careful, we can always add more stuff later.  (I don't think we need to provide this API at all right now...which gives us total freedom to make it do anything we might want in the future...if we provide it now, before we know what we need it for, we are potentially limiting what we can do with it later.)

Reply:
> Why would the UI need `last_name`, `first_name`, or `email`? (I don't think the Dashboard has any more use for those than the Server does....)

On the dashboard User profile page, our UI shows this information to the user, we can remove it from the UI then I guess we wouldn't need this API call but right now the dashboard does show the user's first-name last-name, and email field on the profile page. To get this info, the dashboard must make another OIDC request(userinfo API request) or decode the token itself. On the other hand, Server already has the OIDC token decode mechanism so if the dashboard wants this basic user information It can make a Pbench API request instead of the OIDC request.

Reply:
One thing I can probably do here is I could remove the added admin functionality and only return the decoded user token information when the token is presented. If the user is already an admin and he/she already knows about the username there is little value to get the oidc_id of the target user.

Reply:
> our UI shows this information to the user

That's because, prior to the OIDC integration, the Server required this information in order to create an account, and part of the support for that was to have the UI enable the user to display and update it.

However, now that the Pbench Server doesn't store this information (and doesn't require any of it for login), there is no reason for the Pbench Dashboard to display it to the user or to allow the user to update it -- these are features which the OIDC Identity provider would offer.

So, we should drop this API, unless there is some other reason or requirement for it.  (And, once we have the OIDC integration working, we should remove the vestiges of the local login stuff from the Dashboard.)

Reply:
> However, now that the Pbench Server doesn't store this information (and doesn't require any of it for login), there is no reason for the Pbench Dashboard to display it to the user or to allow the user to update it -- these are features which the OIDC Identity provider would offer.

I agree but that's a dashboard design discussion we might want to have sometime when we have our first-pass OIDC implementation. For example, the dashboard currently displays the `username` at the top-left corner of the UI but unless the dashboard decodes the token or gets the userinfo somehow, it wouldn't know about the username from the token. So not returning the username means breaking this functionality. I am trying to take this forward step by step without breaking the dashboard functionalities. If we decide to remove the profile page from the dashboard we would have to discuss how we want to do that or do we want to replace it with something else. These questions need to be discussed in the design sync. Until that, I am sort of inclined to keep this API call for the dashboard so we don't have as many disruptions on the dashboard code.

Reply:
I'm not sure that the username that the Dashboard is displaying comes from the Pbench Server.  The current mode of operation is for the UI to first get the username _from the user_ and supply it to the server....  (I think what you're seeing comes from a client-side cookie.)

Granted, that will be different when the Dashboard is integrated with OIDC, but it's not obvious to me that the value of displaying the username is worth having a Pbench Server API just to supply it.

And, as for the profile page, we would need to figure out what having one would be _for_.  With the Pbench Server doing its own account maintenance and authentication, the profile page had a purpose; with all of that being handled by OIDC, the profile stuff devolves to Keycloak (unless and until there is some sort of profile which we want the Pbench Server to maintain, which, at the moment, there really isn't).

Reply:
> I'm not sure that the username that the Dashboard is displaying comes from the Pbench Server. The current mode of operation is for the UI to first get the username _from the user_ and supply it to the server.... (I think what you're seeing comes from a client-side cookie.)

Yes exactly and now with the Keycloak integration, we wouldn't have the username at our disposal before the user performs the login and even after that we (Dashboard) have to decode the access token to get the username. So how do we keep the username in the client-side cookie in this case? The OIDC server or the Pbench server would have to let the dashboard know about the username somehow. (One possibility would be to return the username in the payload when the dashboard makes the first Pbsench server API call)


> And, as for the profile page, we would need to figure out what having one would be _for_. With the Pbench Server doing its own account maintenance and authentication, the profile page had a purpose; with all of that being handled by OIDC, the profile stuff devolves to Keycloak (unless and until there is some sort of profile which we want the Pbench Server to maintain, which, at the moment, there really isn't).

I agree that the profile page will not serve its purpose when we integrate fully with OIDC, but until then this API can serve the current needs of the dashboard that's what I am thinking. When we remove the profile page and when we have a design plan to display the user identity (either username or some other information on the UI on the overview page) then we can also remove this API as well.

Reply:
> So how do we keep the username in the client-side cookie in this case?

You mean, how does the Dashboard get the value to put in there?  I assume the answer is going to involve making a significant change to the Dashboard code...at which point all things become possible (and we're no longer talking about a compatibility story). 

> One possibility would be to return the username in the payload when the dashboard makes the first Pbsench server API call

Indeed that's one possibility...and, that's better than maintaining a separate API just for the purpose.  However, having to figure out which call is "the first" strikes me as awkward.  It would make more sense to do something like include that information in the `endpoints` response (i.e., if the request comes with a token, we can reply with the contents; if the token is omitted, then there's nothing to attach), and that's not a bad choice since it's not likely to be a performance-critical one.

> I agree that the profile page will not serve its purpose when we integrate fully with OIDC, but until then this API can serve the current needs of the dashboard that's what I am thinking.

Um, but...what are the "current needs" of the Dashboard during this transition period?  I didn't study it, but I suspect that it will still do what we need it to do if we stop providing the get-user API.  (I hate to say this, but you do know that it is coded to call a `PUT` method, too...right?)

> When we remove the profile page and when we have a design plan to display the user identity (either username or some other information on the UI on the overview page) then we can also remove this API as well.

Given that the profile page will serve no purpose, all that we really have to do is make sure that the Dashboard doesn't crash between now and when we crack it open to add OIDC integration...which may well mean that we can yank the get-user method from the server _now_.

As for user identity, once we switch to OIDC integration, that concept no longer belongs to Pbench (it's provided by the OIDC broker).  If a client like the Dashboard wants to provide "personalization" touches, that really should be between the client and the OIDC broker, although I'm not necessarily against having the Pbench Server provide some assistance if it makes good sense.  But, I feel like that is stuff that we should be adding on later, not trying to provide even before we have OIDC working end-to-end.

Reply:
> You mean, how does the Dashboard get the value to put in there? I assume the answer is going to involve making a significant change to the Dashboard code...at which point all things become possible (and we're no longer talking about a compatibility story).

I am not trying to make it compatible, I am just trying to make the transition is easier on the dashboard.

> Um, but...what are the "current needs" of the Dashboard during this transition period? I didn't study it, but I suspect that it will still do what we need it to do if we stop providing the get-user API. (I hate to say this, but you do know that it is coded to call a `PUT` method, too...right?)

Right now on the dashboard after the successful login, we (the dashboard) immediately make a GET request on the `datasets/list` API call with a bunch of metadata and the `owner` query parameter. We decode the token and check if the requesting `owner` parameter is the same as the one encoded in the token and then we validate the request. When we have the OIDC access token we will follow the same process, we get the token -> decode the token -> check if the requesting `owner` is the same as the one encoded in the token -> validate the user. 

So this process would break on the dashboard because the dashboard has no knowledge of the `username` beforehand. This would need to be fixed on the dashboard and we will have to fix the way we do `check_authorization`. I don't know at the moment what would be the correct mechanism to implement it on the dashboard. The only other way for the dashboard without this API would be to make a direct request to the OIDC server to get the username of the user. And I think its better for the dashboard to hit our API than the OIDC one.

Reply:
And if we really do not want to give any userinfo except the username I could go back to the suggestion you posted at the beginning. The only diff would be instead of making a database call I will still introspect the Bearer token to get the username.

Reply:
And if we all agree that it is okay for the dashboard to hit the OIDC endpoint directly to get the username/user_profile during this transition until we have a clear picture ahead for the dashboard in terms of user profile display or personalization touches I will remove this module. Just note that the [userinfo endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo) is a volatile endpoint and its totally up to the identity provider and broker configuration to expose what info is available on that endpoint. The IDP might even disable the endpoint.

Reply:
> I am not trying to make it compatible, I am just trying to make the transition easier on the dashboard.

😆  You're not trying to make it compatible, you're just trying to keep it from being incompatible??  😀 

> we (the dashboard) immediately make a GET request on the datasets/list API call with a bunch of metadata and the owner query parameter.

So, where does the Dashboard get the value that it uses for the owner query parameter?  I don't _think_ it gets it from the Server, today; so, when the Dashboard is integrated with OIDC, this will be  _new_ problem, not an lack of compatibility.

Moreover, IIRC, the Pbench Server already provides a mechanism for a user to get his/her own data without saying "give me the data for an owner (who is me)", so this new problem may not even be a problem at all.  (Or, if it is indeed a problem, perhaps instead of continuing to require that the Client need to know the user's identity, we could have the Server provide the means for the Client to request "the user's own data" without requiring it to explicitly say who the user is, since that's implicit in the token.)

> if we all agree that it is okay for the dashboard to hit the OIDC endpoint directly to get the username/user_profile

Do remember that, while the Pbench project provides the Dashboard, we won't control all clients of the Pbench Server API.  This means that we should assume that there will be at least one client who hits the Pbench Server `endpoints` API, grabs the OIDC endpoint, and then uses that to the fullest.  And, to the extent that that proves valuable, we might want to consider using it in the Dashboard, too...at some point.

Reply:
> > I am not trying to make it compatible, I am just trying to make the transition easier on the dashboard.
> 
> laughing You're not trying to make it compatible, you're just trying to keep it from being incompatible?? grinning
> 
> > we (the dashboard) immediately make a GET request on the datasets/list API call with a bunch of metadata and the owner query parameter.
> 
> So, where does the Dashboard get the value that it uses for the owner query parameter? I don't _think_ it gets it from the Server, today; so, when the Dashboard is integrated with OIDC, this will be _new_ problem, not an lack of compatibility.

Currently the dashboard gets the username (and password) from the user in order to authenticate with the server. It just remembers that username. With the redirect dance, that's no longer the case as the user only gives their username to Keycloak.

So it's a *new* *compatibility* problem. 😆 

I'm sure there are Javascript packages to decrypt the token directly, but that means the dashboard needs the public key. Possible, but additional mechanism; and being able to ask the server for the decrypted fields makes some sense as a bridge... but I don't feel strongly about it.

> Moreover, IIRC, the Pbench Server already provides a mechanism for a user to get his/her own data without saying "give me the data for an owner (who is me)", so this new problem may not even be a problem at all. (Or, if it is indeed a problem, perhaps instead of continuing to require that the Client need to know the user's identity, we could have the Server provide the means for the Client to request "the user's own data" without requiring it to explicitly say who the user is, since that's implicit in the token.)

`GET /datasets/list` allows you to filter by username or by access. With no filters, it lists everything the authenticated user is allowed to see, which means "mine" and "public". The Overview page doesn't want to show others' public data, so it filters by the authenticated user's username.

We discussed this a few weeks ago; and, yeah, we could add a way to say `?user=<me>`, but we don't have it. (Maybe it'd be `?mine`?) That's certainly possible, and wouldn't be hard, but it's not there now and the question is whether we're going to break the overview page with this PR and then fix it later, or delay this to get in a fix. I don't really see a problem with the server acting as a "decryption helper" in this case, so it's not a big deal.

Reply:
> So it's a new compatibility problem. 😆

OK, so whatever we choose to do, it's going to require changes to the Dashboard and _possibly_ to the Server, as well.

Given that, I don't have a strong leaning between enhancing the `GET /datasets/list` vs. providing a way for the Dashboard to get the user info (of which there are several possibilities).  However, I do feel like retaining the current `GET /user/<target_username>` is a poor choice. Likewise, I don't think providing any new Pbench Server interface specifically for this is a good plan going forward.

So, if we're not going to enhance `/datasets/list`, I suggest enhancing `/endpoints`.

But, neither of these needs to be done in this PR, unless we're _also_ going to add Dashboard changes to this PR, which is think is a little much.

So, I vote that we just delete the `users_api.py` module here and move on.